### PR TITLE
Issue 6344: Add Prometheus metrics support

### LINKED ISSUE #6344: Support Prometheus metrics
**Is your feature request related to a problem? Please describe.**
An operator of a Pravega cluster who uses Prometheus metric tooling can configure Prometheus metrics export from Zookeeper and Bookkeeper, but is unable to do so for Pravega. One possible workaround is to use the Prometheus statsd_exporter, but that's extra deployment complexity that must be managed.

**Describe the solution you'd like**
I'd like to be able to enable Prometheus metrics support from Pravega. The http endpoint that a Prometheus scraper could use could use the existing http endpoint on controller and segment store instances.

**Describe alternatives you've considered**
As noted above, a possible workaround is to use the statsd_exporter. In our configuration, we'd need to manage this as a sidecar kubernetes container to each pravega component container.

### PR DESCRIPTION
**Pull request by alfred-landrum**
**Change log description**  
Adds support for Prometheus metrics as another statistics provider.

**Purpose of the change**  
Closes #6344

**What the code does**  
Allows Prometheus support to be enabled, and a listening port and address to be configured. When the stats provider is started, if Prometheus support is enabled, an http server is started to respond to scrapes.

**How to verify it**  
I added a test to verify that the listening service will be started as configured, and that any added metrics show up in the scraped output. I've also tested locally in standalone mode, and in a Pravega Kubernetes deployment with the Pravega operators.

### CODE REVIEW

File: config/config.properties [hunk 1]
```diff
@@ -299,6 +299,15 @@ metrics.statistics.enable=false
 # InfluxDB retention policy (e.g. 2h, 52w).
 #metrics.influxDB.retention=
 
+# Whether to enable the Prometheus listener.
```
Comment (RaulGracia) [inline config/config.properties:302]: Wondering if we need to update https://github.com/pravega/pravega/blob/master/documentation/src/docs/metrics.md#metrics-service-provider--interface-statsprovider to mention that we also support Prometheus now, right?

**Review by RaulGracia — COMMENTED** [review]
Looks good, thanks a lot. I just left a couple of comments about documentation. One question: has this been tested on a real cluster? I think that from our side, we are going to spin a cluster and check how metrics are being tested.

File: config/config.properties (hunk 1 above)
Comment (alfred-landrum) [reply config/config.properties:302]: done, just added it to the existing phrase in 0489f23

**Comment by alfred-landrum** [comment]
I addressed the existing comments, and made an additional change after more testing: I made the scrape server listen for requests on the pseudo-standard route "/metrics", which is what most prometheus clients use as a default.

I have verified this in a kubernetes cluster, where I've installed Pravega via the pravega operator, and Prometheus via the [kube-prometheus helm chart](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack). Since the controller/segment store pods don't expose a port for the prometheus listener (that's something that the pravega operator could eventually do), it took a little extra work to write the correct monitor object. Here's what worked for me:

```yaml
# Still considering how best to setup monitoring to work with
# a Prometheus operator based setup, but this approach was inspired
# by the conversation on this issue:
# https://github.com/prometheus-operator/prometheus-operator/issues/2515
apiVersion: monitoring.coreos.com/v1
kind: PodMonitor
metadata:
  name: pravega-controller-monitor
spec:
  selector:
    matchLabels:
      app: pravega-cluster
      pravega_cluster: pravega-cluster
      component: pravega-controller
  podMetricsEndpoints:
  # Reference an existing port on the controller,
  # so that Prometheus accepts this as a target.
  - port: grpc
    relabelings:
      # but replace the target address:port with
      # the actual port the Pravega prometheus listener
      # is using:
      - action: replace
        targetLabel: __address__
        sourceLabels:
          - __meta_kubernetes_pod_ip
        replacement: $1:7717
```

**Comment by alfred-landrum** [comment]
Here's screenshot from Grafana showing off metrics from the 2 controllers & 2 segment store pods in my test cluster:
image

**Comment by derekm** [comment]
@alfred-landrum I misunderstood your comment and the context, and CI re-runs are being stubborn about passing... But, yeah, maybe we need to extend that fix to the close() method, as you point out.

**Comment by alfred-landrum** [comment]
I've updated the PR to make the prometheus metrics via the rest server's for both the controller and segmentstore, as @andreipaduroiu requested. I changed the scrape route to `/prometheus` since that port is now shared.

fyi @dellThejas : Here are the kubernetes servicemonitor objects I used to verify this:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: pravega-controller-monitor
spec:
  selector:
    matchLabels:
      app: pravega-cluster
      component: pravega-controller
      pravega_cluster: psi-pvsvc-pv
  endpoints:
    - port: rest
      path: "/prometheus"
 ```

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: pravega-segmentstore-monitor
spec:
  selector:
    matchLabels:
      app: pravega-cluster
      component: pravega-segmentstore
      pravega_cluster: psi-pvsvc-pv
  endpoints:
    - port: server
      path: "/prometheus"
      relabelings:
        - action: replace
          targetLabel: __address__
          sourceLabels:
            - __meta_kubernetes_pod_ip
          replacement: $1:6061
``` 

The segmentstore is different because I don't see a way to enable the segmentstore rest endpoint via the current pravega operator & helm charts. I have to enable that with these options:

```
    pravegaservice.rest.listener.enable: "true"
    pravegaservice.rest.listener.host: "0.0.0.0"
    pravegaservice.rest.listener.host: 6061
```
and then use the relabelings section to target the correct port.

**Comment by alfred-landrum** [comment]
Hi: any further action needed from me to get this merged? The only failing test is the codecov, which as I understand from above, won't report green despite tests that are covering the new code added.

**Comment by andreipaduroiu** [comment]
Yes. You need to add unit tests to cover the new or modified code in the same project where that code lives. This rule is enforced throughout the project and I cannot make exceptions to it.

**Comment by RaulGracia** [comment]
@alfred-landrum it is easy to see the regions of the code that are not yet tested in this diff: https://app.codecov.io/gh/pravega/pravega/compare/6360/diff. If you could add some tests to the new methods added in `StatsProviderImpl` and `RegistryConfigUtil`, I'm sure that the check will pass.

**Comment by alfred-landrum** [comment]
> Yes. You need to add unit tests to cover the new or modified code in the same project where that code lives. This rule is enforced throughout the project and I cannot make exceptions to it.

Ah, ok: from your prior message, I didn't understand that I needed to add tests. Since the tests have been flaky, I misunderstood you, and thought were just describing a limitation of codecov.

**Comment by alfred-landrum** [comment]
> @alfred-landrum it is easy to see the regions of the code that are not yet tested in this diff: https://app.codecov.io/gh/pravega/pravega/compare/6360/diff. If you could add some tests to the new methods added in `StatsProviderImpl` and `RegistryConfigUtil`, I'm sure that the check will pass.

@RaulGracia : thanks for the tip; I added tests within the shared:metrics project.

**Comment by andreipaduroiu** [comment]
This is a great addition @alfred-landrum! Thank you for your contribution.

**Comment by alfred-landrum** [comment]
> This is a great addition @alfred-landrum! Thank you for your contribution.

Thank you very much @andreipaduroiu and @RaulGracia for your help & patience with this! As you may have guessed, my Java experience is low, my background is mostly in Go/C++.

**Pull request merged**